akka-docs/src/test/java/jdocs/stream/operators/sourceorflow/ExtrapolateAndExpandJava.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package jdocs.stream.operators.sourceorflow;
+
+import akka.NotUsed;
+import akka.actor.ActorSystem;
+import akka.actor.Cancellable;
+import akka.japi.Pair;
+import akka.japi.function.Function;
+import akka.stream.javadsl.Flow;
+import akka.stream.javadsl.Sink;
+import akka.stream.javadsl.Source;
+import akka.util.ByteString;
+import docs.stream.operators.sourceorflow.ExtrapolateAndExpandScala;
+import docs.stream.operators.sourceorflow.ExtrapolateAndExpandScala.Frame;
+
+import java.time.Duration;
+import java.util.stream.Stream;
+
+/** */
+public class ExtrapolateAndExpandJava {
+  public static Function<ByteString, Frame> decodeAsFrame =
+      ExtrapolateAndExpandScala.Frame$.MODULE$::decode;
+
+  public static Frame BLACK_FRAME = ExtrapolateAndExpandScala.Frame$.MODULE$.blackFrame();
+
+  public static long nowInSeconds() {
+    return ExtrapolateAndExpandScala.nowInSeconds();
+  }
+
+  public static void main(String[] args) {
+    ActorSystem actorSystem = ActorSystem.create("25fps-stream");
+
+    Source<ByteString, NotUsed> networkSource = ExtrapolateAndExpandScala.networkSource().asJava();
+
+    Flow<ByteString, Frame, NotUsed> decode = Flow.of(ByteString.class).<Frame>map(decodeAsFrame);
+
+    // #extrapolate
+    // if upstream is too slow, produce copies of the last frame but grayed out.
+    Flow<Frame, Frame, NotUsed> rateControl =
+        Flow.of(Frame.class)
+            .extrapolate(
+                lastFrame -> {
+                  Frame gray =
+                      new Frame(
+                          ByteString.fromString(
+                              "gray frame!! - " + lastFrame.pixels().utf8String()));
+                  return Stream.iterate(gray, i -> i).iterator();
+                },
+                BLACK_FRAME // initial value
+                );
+
+    Source<Frame, NotUsed> videoSource = networkSource.via(decode).via(rateControl);
+
+    // let's create a 25fps stream (a Frame every 40.millis)
+    Source<String, Cancellable> tickSource =
+        Source.tick(Duration.ZERO, Duration.ofMillis(40), "tick");
+
+    Source<Frame, Cancellable> videoAt25Fps = tickSource.zip(videoSource).map(Pair::second);
```

Comment:
Hmm, perhaps instead of zipping with a `tickSource` we should zip with the audio stream? (though then we should also make sure to somehow drop late frames... it's OK if these examples are a bit contrived I guess)

Reply:
There's all sorts of timing issues trying to align video and audio and I just wanted to demo a slo source (network) with a consumers that has a very clear demand (25 elements per second). :-)